src/chat.rs

```diff
@@ -362,6 +359,57 @@ impl ChatId {
             .unwrap_or_default() as usize
     }
 
+    pub(crate) fn get_param(self, context: &Context) -> Result<Params, Error> {
+        let res: Option<String> = context
+            .sql
+            .query_get_value_result::<_, _>("SELECT param FROM chats WHERE id=?", params![self])?;
+        Ok(res
+            .map(|s| s.parse().unwrap_or_default())
+            .unwrap_or_default())
+    }
+
+    // Returns true if chat is a saved messages chat.
+    pub fn is_self_talk(self, context: &Context) -> Result<bool, Error> {
+        Ok(self.get_param(context)?.exists(Param::Selftalk))
+    }
+
+    /// Returns true if chat is a device chat.
+    pub fn is_device_talk(self, context: &Context) -> Result<bool, Error> {
+        Ok(self.get_param(context)?.exists(Param::Devicetalk))
+    }
+
+    /// Hides or deletes messages which are expired according to
+    /// "delete_device_after" setting.
+    ///
+    /// Returns true if any message is hidden, so event can be emitted. If
+    /// nothing has been hidden, returns false.
+    pub fn delete_device_expired_messages(self, context: &Context) -> Result<bool, Error> {
+        if self.is_special() || self.is_self_talk(context)? || self.is_device_talk(context)? {
+            return Ok(false);
+        }
+
+        if let Some(delete_device_after) = context.get_config_delete_device_after() {
+            let threshold_timestamp = time() - delete_device_after;
+
+            // Hide expired messages
+            //
+            // Only update the rows that have to be updated, to avoid emitting
+            // unnecessary "chat modified" events.
+            let rows_modified = context.sql.execute(
+                "UPDATE msgs \
+                 SET txt = 'DELETED', hidden = 1 \
+                 WHERE timestamp < ? \
+                 AND chat_id == ? \
```

Comment:
i would just check `chat_id > DC_CHAT_ID_LAST_SPECIAL` here and make delete_device_expired_messages() part of context, not of chat/chat_id. this makes `delete_device_expired_messages_all_chats()` superfluous and dc_get_chatlist() and similar functions much faster as it does not have to call delete_device_expired_messages() 500 times (eg. on my delta)

Reply:
`chat_id > DC_CHAT_ID_LAST_SPECIAL` is not enough to avoid deleting messages from "saved messages" chat and device chat.

Reply:
refactored into a single function